overlord/snapstate/autorefresh_gating.go

```diff
@@ -32,6 +35,289 @@ import (
 
 var gateAutoRefreshHookName = "gate-auto-refresh"
 
+// cumulative hold time for snaps other than self
+const maxOtherHoldDuration = time.Hour * 48
+
+var timeNow = func() time.Time {
+	return time.Now()
+}
+
+func lastRefreshed(st *state.State, snapName string) (time.Time, error) {
+	var snapst SnapState
+	if err := Get(st, snapName, &snapst); err != nil {
+		return time.Time{}, fmt.Errorf("internal error, cannot get snap %q: %v", snapName, err)
+	}
+	// try to get last refresh time from snapstate, but it may not be present
+	// for snaps installed before the introduction of last-refresh attribute.
+	if snapst.LastRefreshTime != nil {
+		return *snapst.LastRefreshTime, nil
+	}
+	snapInfo, err := snapst.CurrentInfo()
+	if err != nil {
+		return time.Time{}, err
+	}
+	// fall back to the modification time of .snap blob file as it's the best
+	// approximation of last refresh time.
+	fst, err := os.Stat(snapInfo.MountFile())
+	if err != nil {
+		return time.Time{}, err
+	}
+	return fst.ModTime(), nil
+}
+
+type holdInfo struct {
+	// FirstHeld keeps the time when the given snap was first held for refresh by a gating snap.
+	FirstHeld time.Time `json:"first-held"`
+	// HoldUntil stores the desired end time for holding.
+	HoldUntil time.Time `json:"hold-until"`
+}
+
+func refreshGating(st *state.State) (map[string]map[string]*holdInfo, error) {
+	// held snaps -> holding snap(s) -> first-held/hold-until time
+	var gating map[string]map[string]*holdInfo
+	err := st.Get("snaps-hold", &gating)
+	if err != nil && err != state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot get snaps-hold: %v", err)
+	}
+	if err == state.ErrNoState {
+		return make(map[string]map[string]*holdInfo), nil
+	}
+	return gating, nil
+}
+
+// HoldError contains the details of snaps that cannot to be held.
+type HoldError struct {
+	SnapsInError map[string]error
```

Comment:
I think for each of the snaps with a problem we want to compute the maximum allowed hold left if any and also store it in the error. This is the minimum  (maxDur+firstHeld-now),  (maxPostponement+lastRefresh-now) I think?